features/64-chat-offline-cache/api.md

```diff
@@ -0,0 +1,318 @@
+```
+
+#### JS
+
+```js
+const successCallback = (conversations, cached = false) => {
+  // display conversations
+};
+
+skygearChat
+  .getConversations(successCallback)
+  .then(successCallback, errorCallback);
+
+// no-cache version
+skygearChat
+  .getConversations()
+  .then(successCallback, errorCallback);
+```
+
+```js
+let currentMessages = [];
+
+const successCallback = (messages, cached = false) => {
+  currentMessages = skygearChat.merge(currentMessages, messages);
+  // display currentMessages
+};
+
+skygearChat
+  .getMessages(conversation, beforeMessage, limit, order, successCallback)
+  .then(successCallback, errorCallback);
+
+// no-cache version
+skygearChat
+  .getMessages(conversation, beforeMessage, limit, order)
+  .then(successCallback, errorCallback);
+```
+
+#### Non-JS
+
+```
+skygearChat.fetchConversations(func (conversations, cached = false) {
+  // display conversations
+})
+```
+
+```
+let currentMessages = [];
+
+skygearChat.fetchMessages(func (messages, cached = false) {
+  currentMessages = skygearChat.merge(currentMessages, messages);
+  // display currentMessages
+})
+```
+
+- Add a managed message list for user to use directly
+
+```
+class ManagedMessageList {
+
+  constructor(container: SkygearContainer,
+              conversation: Conversation,
+              delegate: ManagedMessageListDelegate)
+
+  - fetchLimit: Int = 100
+  - useCache: Boolean = true
+  - syncOnReconnect: Boolean = true
+
+  func startSubscribing() // expected to call when enter view
+
+  func stopSubscribing() // expected to call when leave view
+
+  func messageCount() -> Int
+
+  func messageAt(position: Int) -> Message
+
+  func fetchLatestMessages()
+
+  protected func handleFetchedMessages(messages: [Message]) -> Void
+  protected func handleCachedMessages(messages: [Message]) -> Void
+
+  // handle pubsub change
+  protected func handleMessageChanges(messageDeltas: [MessageDelta]) -> Void
+
+  // for UI display
+  - protected messages: [Message]
+
+  // to determine when to fetch messages, may need to consider gap between messages
+  // consider a client app may stay on previous message position
+  // and have a skip to latest button
+  // while the current position and the latest message has a gap larger than fetch limit
+  //
+  // if useCache = false
+  // messages == fetchedMessages
+  - protected fetchedMessages: [Message]
+
+  /**
+   * scrollViewDelegateMethods
+   *
+   * the list will determine when to fetch messages by checking current first
+   * and last visible message
+   *
+   * the implementation would call fetchMessages automatically
+   */
+
+}
+
+interface ManagedMessageListDelegate {
+
+  func managedMessageListDidChange(messageDeltas: [MessageDelta])
+
+}
+```
+
+- Add cache store to skygear chat
+
+There should be a common cache logic layer and cache implementation.
+
+#### Cache implementation
+
+Skygear may provide cache implementation for each platform.
+
+- iOS: Core Data
```

Comment:
Is Core Data fast enough now? Because I heard from a layer user, he said Layer's team said their original implementation use Core Data and was very very slow.